overlord/assertstate/assertmgr.go

```diff
@@ -86,3 +105,182 @@ func Add(s *state.State, a asserts.Assertion) error {
 	// TODO: deal together with asserts itself with (cascading) side effects of possible assertion updates
 	return cachedDB(s).Add(a)
 }
+
+// TODO: snapstate also has this, move to auth, or change a bit the approach now that we have AuthContext in the store?
+func userFromUserID(st *state.State, userID int) (*auth.UserState, error) {
+	if userID == 0 {
+		return nil, nil
+	}
+	return auth.User(st, userID)
+}
+
+// fetch fetches or updates the referenced assertion and all its prerequisites from the store and adds them to the system assertion database. It does not fail if required assertions were already present.
+func fetch(s *state.State, ref *asserts.Ref, userID int) error {
+	// TODO: once we have a bulk assertion retrieval endpoint this approach will change
+
+	user, err := userFromUserID(s, userID)
+	if err != nil {
+		return err
+	}
+
+	db := cachedDB(s)
+	store := snapstate.Store(s)
+
+	s.Unlock()
+	defer s.Lock()
+
+	got := ([]asserts.Assertion)(nil)
+
+	retrieve := func(ref *asserts.Ref) (asserts.Assertion, error) {
+		// TODO: ignore errors if already in db?
+		return store.Assertion(ref.Type, ref.PrimaryKey, user)
+	}
+
+	save := func(a asserts.Assertion) error {
+		got = append(got, a)
+		return nil
+	}
+
+	f := &fetcher{
+		retrieve: retrieve,
+		save:     save,
+	}
+	f.init(db)
+
+	err = f.doFetch(ref)
+	if err != nil {
+		return err
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	// TODO: deal together with asserts itself with (cascading) side effects of possible assertion updates
+
+	for _, a := range got {
+		err := db.Add(a)
+		if revErr, ok := err.(*asserts.RevisionError); ok {
+			if revErr.Current >= a.Revision() {
```

Comment:
I don't understand the question, there is it's flagger as an error, here we want to ignore it such errors to be idempotent and also because we try to update some assertions that are shared between many snaps so this happens naturally